### PR TITLE
BCDA-1506: Create auth/system/:id:/credentials PUT endpoint

### PR DESCRIPTION
**Pull request by embh**
### Fixes [BCDA-1506](https://jira.cms.gov/browse/BCDA-1506)
The SSAS needs a mechanism for resetting credentials.

### Proposed changes:
Create `PUT /system/[id]/credentials`, which tells the service to disable the current secret for the system and create a new one. The new secret is returned.

### Change Details
* Creates router `PUT /system/[id]/credentials` path
* Adds `GetSystemByID()` and `ResetSecret()` to systems.go
* Adds `resetCredentials()` to admin/api.go

### Security Implications
This change resets the secret for a given system, which can help improve security.

### Acceptance Validation
Old and new tests pass.

### Feedback Requested
Any. A few questions:
* Should the logged event be called `ResetSecret` or `NewSecret`? My aim was to distinguish it from when a secret is created for the first time, but that might not be necessary.
* Is the response body format correct? Changes to content and/or naming? `{ "client_id": "1234", "client_secret": "secret" }`

### CODE REVIEW

File: ssas/systems.go [hunk 1]
```diff
@@ -414,6 +431,41 @@ func GenerateSecret() (string, error) {
...
+func (system *System) ResetSecret(trackingID string) (string, error) {
+	db := GetGORMDbConnection()
+	defer Close(db)
+
+	newSecretEvent := Event{Op: "ResetSecret", TrackingID: trackingID, ClientID: system.ClientID}
+	OperationStarted(newSecretEvent)
+
+	secretString, err := GenerateSecret()
+	if err != nil {
+		newSecretEvent.Help = fmt.Sprintf("could not reset secret for clientID %s: %s", system.ClientID, err.Error())
```
Comment (dhgreene) [inline ssas/systems.go]: Presumably an error here or elsewhere in this function would have a message clear enough to distinguish what failed.  Do you think those that follow us might ever ask, "Did that unique cryptographic error bubble up from `GenerateSecret()` or `NewHash()`?"  I've often come down on the side of generating more unique error messages to combat that possibility, but what you have here has the upside of consistency.

File: ssas/systems.go [hunk 2]
```diff
@@ -414,6 +431,41 @@ func GenerateSecret() (string, error) {
...
+		OperationFailed(newSecretEvent)
+		return "", errors.New("internal system error")
```
Comment (dhgreene) [inline ssas/systems.go]: 👍 on returning a more generic error message, so that future public-facing endpoints can safely pass along the message.

**Comment by dhgreene** [comment]
I like the unique name for the logged event.  The response body is great: it parallels the output from self-registration (which is specified by the RFC), but omits the unneeded system name.

File: ssas/systems.go (hunk 1 above)
Comment (DeirdreHolub) [reply ssas/systems.go]: The GenerateSecret() and NewHash() methods log their operation start, failure and success, so there should be no wonder here. In contrast to most approaches with error handling, which take a top down approach to logging (log the error at the last place it's handled), we are logging from the bottom up. We're doing this to provide useful diagnostic logs in production and also useful warning metrics.

**Pull request merged**